plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/panel/GitPanelPresenter.java

```diff
@@ -0,0 +1,360 @@
+          if (projectName.equals(selectedProjectName)) {
+            updateChangedFiles(alteredFiles);
+          }
+        }
+        break;
+      default:
+        // do nothing
+    }
+  }
+
+  /** Handles creation and deletion of projects. */
+  @Override
+  public void onResourceChanged(ResourceChangedEvent event) {
+    ResourceDelta delta = event.getDelta();
+    Resource resource = delta.getResource();
+    if (resource.isProject() && resource.getLocation().segmentCount() == 1) {
+      // resource is a root project
+      if (isProjectUnderGit(resource.asProject())) {
+        if (delta.getKind() == ResourceDelta.ADDED) {
+          if ((delta.getFlags() & (MOVED_FROM | MOVED_TO)) != 0) {
+            // project renamed
+            String oldProjectName = delta.getFromPath().segment(0);
+            String newProjectName = delta.getToPath().segment(0);
+            MutableAlteredFiles alteredFiles =
+                new MutableAlteredFiles(
+                    findProjectByName(newProjectName), changes.remove(oldProjectName));
+
+            changes.put(newProjectName, alteredFiles);
+            view.renameRepository(oldProjectName, newProjectName);
+          } else {
+            // project created
+            changes.put(resource.getName(), new MutableAlteredFiles(resource.asProject()));
+            view.addRepository(resource.getName());
+          }
+        } else if (delta.getKind() == ResourceDelta.REMOVED) {
+          // project deleted
+          changes.remove(resource.getName());
+          view.removeRepository(resource.getName());
+        }
+      }
+    } else {
+      // handle deletion of an item
+      if (delta.getKind() == ResourceDelta.REMOVED) {
+        String pathToItem = delta.getResource().getLocation().toString();
+        String projectName = extractProjectName(pathToItem);
+        MutableAlteredFiles alteredFiles = changes.get(projectName);
+        if (alteredFiles == null) {
+          // project doesn't have a git repository
+          return;
+        }
+        String itemRelativePath = removeProjectName(pathToItem);
+
+        if (alteredFiles.getStatusByFilePath(itemRelativePath) != Status.ADDED) {
+          alteredFiles.addFile(itemRelativePath, Status.DELETED);
+        } else {
+          alteredFiles.removeFile(itemRelativePath);
+        }
+
+        view.updateRepositoryChanges(projectName, alteredFiles.getFilesQuantity());
+        if (projectName.equals(selectedProjectName)) {
+          updateChangedFiles(alteredFiles);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void onGitRepositoryInitialized(
+      String endpointId, RepositoryInitializedEventDto repositoryInitializedEventDto) {
+    String projectName = repositoryInitializedEventDto.getProjectName();
+    changes.put(projectName, new MutableAlteredFiles(findProjectByName(projectName)));
+    view.addRepository(projectName);
+  }
+
+  @Override
+  public void onGitRepositoryDeleted(
```

Comment:
`selectedProjectName` might refer to unexisted project

Reply:
It will be changed in `onRepositorySelectionChanged`. But you are right, let's get rid of this variable, it's error prone.